python/lsst/pipe/tasks/makeGen3SkyMap.py

```diff
@@ -0,0 +1,119 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+import lsst.afw.geom as afwGeom
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+from lsst.daf.butler import Butler, DatasetType
+from lsst.skymap import skyMapRegistry
+
+
+class MakeGen3SkyMapConfig(pexConfig.Config):
+    """Config for MakeGen3SkyMapTask
+    """
+    skyMapName = pexConfig.Field(
+        doc="Name assigned to created skymap in butler registry",
+        dtype=str,
+        default="deepCoadd_skyMap",
+    )
+    skyMap = skyMapRegistry.makeField(
+        doc="type of skyMap",
+        default="dodeca",
+    )
+    doWrite = pexConfig.Field(
+        doc="persist the skyMap? If False then run generates the sky map and returns it, "
+            "but does not save it to the data repository",
+        dtype=bool,
+        default=True,
+    )
+
+
+class MakeGen3SkyMapTask(pipeBase.Task):
+    ConfigClass = MakeGen3SkyMapConfig
+    _DefaultName = "makeGen3SkyMap"
+
+    """This is a task to construct and optionally save a SkyMap into a gen3
+    butler repository.
+
+    Parameters
+    ----------
+    config : `MakeGen3SkyMapConfig` or None
+        Instance of a configuration class specifying task options, a default
+        config is created if value is None
+    butler : `str`
+        Path to a gen3 butler
+    collection : `str`
+        The name of the collection which the created skymap will be inserted
+    """
+
+    def __init__(self, *, config=None, butler=None, collection=None, **kwargs):
+        if butler is None:
+            raise ValueError("A path to a butler must be given")
+        if collection is None:
+            raise ValueError("The collection to use with this butler must be provided")
+        super().__init__(config=config, **kwargs)
+        self.butler = Butler(butler, run=collection)
+        self.collection = collection
+
+    def run(self):
+        """Construct and optionally save a SkyMap into a gen3 repository"""
+        skyMap = self.config.skyMap.apply()
+        self.logSkyMapInfo(skyMap)
+        skyMapHash = skyMap.getSha1()
+        if self.config.doWrite:
+            try:
+                existing, = self.butler.registry.query("SELECT skymap FROM skymap WHERE hash=:hash",
+                                                       hash=skyMapHash)
+                raise RuntimeError(
+                    (f"SkyMap with name {existing.name} and hash {skyMapHash} already exist in "
+                     f"the butler collection {self.collection}, SkyMaps must be unique within "
+                     "a collection")
+                )
+            except ValueError:
+                self.log.info(f"Inserting SkyMap {self.config.skyMapName} with hash={skyMapHash}")
+                with self.butler.registry.transaction():
+                    skyMap.register(self.config.skyMapName, self.butler.registry)
```

Comment:
It seems I'm not able to create one deepCoadd_skyMap  skymap with config1 and then create another deepCoadd_skyMap  skymap with config2, even with a different butler collection name. I would get `lsst.daf.butler.core.registry.ConflictingDefinitionError: Existing definition for skymap entry with {skymap: deepCoadd_skyMap}.` 

Is this intentional? Is the idea to change the skyMapName when a different config is used?

Reply:
Hmm... it seems changing `skyMapName` doesn't change the hash; the hash is determined by other config parameters. 

Is it not allowed to make two skymaps to two collections using the same configs, for example, 
`makeGen3Skymap.py REPO collection1`  and then `makeGen3Skymap.py REPO collection2` ?

Reply:
In today's standup @TallJimbo clarified that a duplicate skymap of the same set of configs (determining the hash; `skyMapName` not included) cannot be ingested no matter the collections. That explains the `ConflictingDefinitionError` I saw. It seems okay to me for now, when we each have our own sqlite3 registry. I'm not sure how it will be when we share things on one Oracle database, but probably fine for now.